internal/core/textlayout.rs

```diff
@@ -0,0 +1,954 @@
+        &TextLine<Font::Length>,
+    ),
+) -> Font::Length {
+    let wrap = wrap == TextWrap::word_wrap;
+    let elide_glyph =
+        if overflow == TextOverflow::elide { font.glyph_for_char('…') } else { None };
+    let max_width_without_elision =
+        max_width - elide_glyph.as_ref().map_or(Font::Length::zero(), |g| font.glyph_advance_x(g));
+
+    let glyphs = RefCell::new(Vec::new());
+
+    let new_line_break_iter = |glyphs| {
+        TextLineBreaker::new(string, font, glyphs, if wrap { Some(max_width) } else { None })
+    };
+    let mut text_lines = None;
+
+    let mut text_height = |glyphs| {
+        if single_line {
+            font_height
+        } else {
+            text_lines = Some(new_line_break_iter(glyphs).collect::<Vec<_>>());
+            font_height * (text_lines.as_ref().unwrap().len() as i16).into()
+        }
+    };
+
+    let two = Font::LengthPrimitive::one() + Font::LengthPrimitive::one();
+
+    let baseline_y = match vertical_alignment {
+        TextVerticalAlignment::top => Font::Length::zero(),
+        TextVerticalAlignment::center => max_height / two - text_height(&glyphs) / two,
+        TextVerticalAlignment::bottom => max_height - text_height(&glyphs),
+    };
+
+    let mut y = baseline_y;
+
+    let mut process_line =
+        |line: &TextLine<Font::Length>, glyphs: &RefCell<Vec<(Font::Glyph, usize)>>| {
+            let x = match horizontal_alignment {
+                TextHorizontalAlignment::left => Font::Length::zero(),
+                TextHorizontalAlignment::center => {
+                    max_width / two - euclid::approxord::min(max_width, line.text_width) / two
+                }
+                TextHorizontalAlignment::right => {
+                    max_width - euclid::approxord::min(max_width, line.text_width)
+                }
+            };
+
+            let mut elide_glyph = elide_glyph.as_ref().clone();
+
+            let glyphs = glyphs.borrow();
+            let glyph_it = glyphs[line.glyph_range.clone()].iter();
+            let mut glyph_x = Font::Length::zero();
+            let mut positioned_glyph_it = glyph_it.filter_map(|(glyph, _)| {
+                // TODO: cut off at grapheme boundaries
+                if glyph_x >= max_width_without_elision {
+                    if let Some(elide_glyph) = elide_glyph.take() {
+                        return Some((glyph_x, elide_glyph));
+                    } else {
+                        return None;
+                    }
+                }
+                let positioned_glyph = (glyph_x, glyph);
+                glyph_x += font.glyph_advance_x(glyph);
+                Some(positioned_glyph)
+            });
+
+            layout_line(&mut positioned_glyph_it, x, y, line);
+            y += font_height;
+        };
+
+    if let Some(lines_vec) = text_lines.take() {
+        for line in lines_vec {
+            process_line(&line, &glyphs);
+        }
+    } else {
+        let glyphs = RefCell::new(Vec::new());
```

Comment:
```suggestion
```

(I think you can just re-use the other one since it should still be empty)

Reply:
Ah yes. I tried that earlier but the borrow checker complained. Now that it's in a RefCell this might be possible. I'll try :)